fix(service-worker): treat 503 as offline

alexwiese opened the pull request:
Prior to this commit the service worker only treated 504 errors as "effectively offline".
This commit changes the behaviour to treat both 503 (Service Unavailable) and 504 as "offline".

Fixes #35571

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #35571


## What is the new behavior?

503 errors will be treated as "offline" and the cached content will be served.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alexwiese commented:
@googlebot I signed it!

alexwiese commented:
Looks like the build/tests failed due to something outside of this PR. How can I trigger the builds again?

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -684,7 +684,7 @@ export class Driver implements Debuggable, UpdateSource {
    * Retrieve a copy of the latest manifest from the server.
    * Return `null` if `ignoreOfflineError` is true (default: false) and the server or client are
-   * offline (detected as response status 504).
+   * offline (detected as response status 504 or 503).
```
I think it would be helpful to include the status test in parenthesis:

```suggestion
   * offline (detected as response status 503 (service unavailable) or 504 (gateway timeout)).
```

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -695,7 +695,7 @@ export class Driver implements Debuggable, UpdateSource {
         await this.deleteAllCaches();
         await this.scope.registration.unregister();
-      } else if (res.status === 504 && ignoreOfflineError) {
+      } else if ((res.status === 504 || res.status === 503) && ignoreOfflineError) {
```
```suggestion
      } else if ((res.status === 503 || res.status === 504) && ignoreOfflineError) {
```

:grin:

gkalpak reviewed (changes requested):
Thx, @alexwiese :+1:

Could you also add a test for this. E.g. something similar to [this test][1]. To do that, we need a way to tell the mock server to return a 503 response. (We could use something similar to the [`online` flag][2] or (since this is a one-off thing) spy on `server.fetch()` and return `new MockResponse(null, {status: 503, statusText: 'Service Unavailable'})`).

[1]: https://github.com/angular/angular/blob/e084835fb/packages/service-worker/worker/test/happy_spec.ts#L864
[2]: https://github.com/angular/angular/blob/e084835fb/packages/service-worker/worker/testing/mock.ts#L106

alexwiese committed:
fix(service-worker): treat 503 as offline

Prior to this commit the service worker only treated 504 errors as "effectively offline".
This commit changes the behaviour to treat both 503 (Service Unavailable) and 504 as "offline".

Fixes #35571

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -872,6 +872,21 @@ import {SwTestHarness, SwTestHarnessBuilder} from '../testing/scope';
+         expect(driver.state).toEqual(DriverReadyState.NORMAL);
+         expect(scope.unregistered).toBeFalsy();
+         expect(await scope.caches.keys()).not.toEqual([]);
+       });
```
Thx for adding the test. Have you verified that it fails without the fix?

alexwiese replied:
I can’t get the tests to run properly on my machine. I get a folder/file not found error. Will try on another system.

gkalpak replied:
How are you trying to run them?

alexwiese replied:
```git clone https://github.com/alexwiese/angular/
cd angular
yarn install
node ./scripts/build-packages-dist.js
yarn bazel test packages/service-worker```
The error is  `C:/users/alex/_bazel_alex/3tbqurya/execroot/angular/bazel-out/host/bin/external/npm/@bazel/typescript/bin/tsc_wrapped.sh: No such file or directory`

alexwiese replied:
OK I spun up a brand new Ubuntu VM, installed node 12.16.1, yarn 1.21.1, ran the same commands and I get a similar error:

> error trying to exec 'cc1plus': execvp: No such file or directory

alexwiese replied:
OK installing g++ fixed it. Not sure if that indicates an issue with the build scripts or the instructions.
`sudo apt-get install g++`
Once I did that on the Ubuntu VM I was able to build and test.
I have confirmed the test fails without the fix, and passes with the fix.

gkalpak replied:
Thanks for figuring this out, @alexwiese :+1:

BTW, you shouldn't need `node ./scripts/build-packages-dist.js`. The `test` command should be a ble to build the necessary packages.

`g++` is indeed one of the required dependencies of bazel (see [here][1]).

(No idea where the other error came from - I've never seen that 😕)

[1]: https://docs.bazel.build/versions/master/install-ubuntu.html#step-1-install-required-packages
